src/firejail/fs.c

```diff
@@ -96,11 +96,20 @@ static void disable_file(OPERATION op, const char *filename) {
 
 	// if the file is not present, do nothing
 	assert(fname);
+	int fd = open(fname, O_PATH|O_CLOEXEC);
```

Comment:
@smitsohu [on Oct 29](https://github.com/netblue30/firejail/pull/4652#discussion_r739130414):

> The idea is cool, but I have few remarks: You accumulate open file
> descriptors if there are `read-only`, `read-write`, `noexec` or `tmpfs`
> options.

I don't understand what you mean.  Do the flags used affect the total number of
file descriptors?  Isn't every file descriptor going to use the same amount of
memory and count towards the limit?

Also, in this case only one fd is used, so I don't get it.  Do you mean to use
`O_RDONLY` instead of `O_PATH` or maybe to use no flags at all?

> Related, it is probably a good idea to not run big chunks of code with files
> unnecessarily open.

Agreed.

> I think you can resolve both issues by rearranging the code a bit

The only obvious improvement that I can see would be to move the `if
(strcmp(BINDIR "/firejail", fname) == 0) {` section to above this line.

> and closing the file descriptor right after blacklist processing.

It is currently closed after `EUID_USER();` as it is used right above it, so
I'm not sure what else could be done (at least not without changing the
behavior).

Reply:
Files are opened but not closed if we fall further down the if-else chain (op == MOUNT_READONLY || op == MOUNT_RDWR || op == MOUNT_NOEXEC || op == MOUNT_TMPFS).

>  improvement that I can see would be to move the if (strcmp(BINDIR "/firejail", fname) == 0) { section to above this line.

Agreed.

> Do you mean to use O_RDONLY instead of O_PATH

Obviously there is not always a choice, but in general `O_PATH` is better. Sometimes opening paths with `O_RDONLY` can be enough to create unwanted side effects, like for example Firejail could become the only reader on a FIFO.

Reply:
By the way, I think that once your improvement has landed, it will become an irresistible temptation to create a version of fs_tmpfs that accepts file descriptors as argument, and use your new file descriptor in the `op == MOUNT_TMPFS` case, too.

Reply:
@smitsohu [on Oct 30](https://github.com/netblue30/firejail/pull/4652#discussion_r739636058):

> Files are opened but not closed if we fall further down the if-else chain (op
> == MOUNT_READONLY || op == MOUNT_RDWR || op == MOUNT_NOEXEC || op ==
> MOUNT_TMPFS).

Good catch; fixed.

> > improvement that I can see would be to move the if (strcmp(BINDIR
> > "/firejail", fname) == 0) { section to above this line.
> 
> Agreed.

Done.

> > Do you mean to use O_RDONLY instead of O_PATH
> 
> Obviously there is not always a choice, but in general `O_PATH` is better.
> Sometimes opening paths with `O_RDONLY` can be enough to create unwanted side
> effects, like for example Firejail could become the only reader on a FIFO.
> Also it is faster.

+1

@smitsohu [on Oct 30](https://github.com/netblue30/firejail/pull/4652#discussion_r739641807):

> By the way, I think that once your improvement has landed, it will become an
> irresistible temptation to create a version of fs_tmpfs that accepts file
> descriptors as argument, and use your new file descriptor in the `op ==
> MOUNT_TMPFS` case, too.

Indeed; I noticed that on other functions on other TOCTOU-related commits, that
actually fixing potential race conditions (besides just the TOCTOU warnings)
would probably require certain functions that take paths to also have a version
that takes fds (I managed to create one of those for a very simple function,
`set_perms`).  Though one obstacle that I've seen is when some of them also
need to test for symlinks, but at least it should be simpler for the other
ones.

Reply:
The question is also to what end we want to remove these potential races.

If a function is good enough for what it does (because it operates on paths that obviously are not under control of a user), and the fix to a race condition does not generally improve the quality of the code (for example because things are getting much more complicated), then it is probably as good or even better to just add a comment and turn off the warning. But that's only IMHO.

Reply:
@smitsohu [on Oct 31](https://github.com/netblue30/firejail/pull/4652#discussion_r739806713):

> The question is also to what end we want to remove these potential races.
> 
> If a function is good enough for what it does (because it operates on paths
> that obviously are not under control of a user), and the fix to a race
> condition does not generally improve the quality of the code (for example
> because things are getting much more complicated), then it is probably as
> good or even better to just add a comment and turn off the warning. But
> that's only IMHO.

Yes, I don't think every single instance has to be changed.  It can get rather
verbose with the symlink tests (and for functions with no fd equivalent, such
as `unlink`/`unlinkat`), as a dirfd would likely be needed besides the normal
fd.  Though in most of the WIP commits related to these warnings, the code ends
up looking similar enough or IMO looking a bit better due to some small
refactorings.

I'd also note that some helper functions (e.g.: on util.c) that might only be
used for read-only paths now could be used for read-write paths later, so I
think that those would generally seem to be good targets for minimizing race
conditions.